src/compose/service-manager.coffee

```diff
@@ -0,0 +1,276 @@
+Promise = require 'bluebird'
+_ = require 'lodash'
+EventEmitter = require 'events'
+JSONParser = require 'jsonparse'
+fs = Promise.promisifyAll(require('fs'))
+
+logTypes = require '../lib/log-types'
+{ checkInt } = require '../lib/validation'
+constants = require '../lib/constants'
+
+Service = require './service'
+#restartVars = (conf) ->
+#	return _.pick(conf, [ 'RESIN_DEVICE_RESTART', 'RESIN_RESTART' ])
+
+module.exports = class ServiceManager extends EventEmitter
+	constructor: ({ @docker, @logger, @config }) ->
+		@containerHasDied = {}
+		@listening = false
+		# Volatile state of containers, indexed by containerId (or random strings if we don't have a containerId yet)
+		@volatileState = {}
+
+	killAllLegacy: =>
+		# Containers haven't been normalized (this is an updated supervisor)
+		# so we need to stop and remove them
+		@docker.getImage(constants.supervisorImage).inspect()
+		.then (supervisorImage) =>
+			Promise.map(@docker.listContainers(all: true), (container) =>
+				@docker.getImage(container.Image).inspect()
+				.then (containerImage) =>
+					if containerImage.Id != supervisorImage.Id
+						@_killContainer(container.Id, { serviceName: 'legacy', image: container.Image }, { removeContainer: true })
+			)
+
+	reportChange: (containerId, status) ->
+		if status?
+			@volatileState[containerId] ?= {}
+			_.merge(@volatileState[containerId], status)
+			@emit('change')
+		else if containerId? and @volatileState[containerId]?
+			delete @volatileState[containerId]
+			@emit('change')
+
+	reportNewStatus: (containerId, service, status) =>
+		@reportChange(containerId, _.merge({ status }, _.pick(service, [ 'imageId', 'appId', 'releaseId', 'commit' ])))
+
+	_killContainer: (containerId, service = {}, { removeContainer = true }) =>
+		@logger.logSystemEvent(logTypes.stopService, { service })
+		@reportNewStatus(containerId, service, 'Stopping') if service.imageId?
+		containerObj = @docker.getContainer(containerId)
+		containerObj.stop(t: 10)
+		.then ->
+			containerObj.remove(v: true) if removeContainer
+		.catch (err) =>
+			# Get the statusCode from the original cause and make sure statusCode its definitely a string for comparison
+			# reasons.
+			statusCode = '' + err.statusCode
```

Comment:
Why not make it a number since you compare it with numbers?

Reply:
Done (it had been like that before my time, I suppose because dockerode might have sometimes returned numbers or strings? I added a checkInt just in case)